io_scene_godot/converters/mesh.py

```diff
@@ -151,17 +159,126 @@ def make_arrays(escn_file, export_settings, node, armature_data):
+        for vert_data in surface.morph_arrays:
+            vert_data.has_bone = has_bone
 
     bpy.data.meshes.remove(mesh)
 
+    if armature_data is not None:
+        armature_data.pose_position = original_pose_position
+        bpy.context.scene.update()
+
     return surfaces
 
 
+def extract_shape_keys(blender_shape_keys):
+    """Return a list of (shape_key_index, shape_key_object) each of them
+    is a shape key needs exported"""
+    # base shape key needn't be exported
+    ret = list()
+    base_key = blender_shape_keys.reference_key
+    for index, shape_key in enumerate(blender_shape_keys.key_blocks):
+        if shape_key != base_key:
+            ret.append((index, shape_key))
+    return ret
+
+
+def export_morphs(export_settings, node, surfaces, has_tangents,
+                  gid_to_bid_map):
+    """Export shape keys in mesh node and append them to surfaces"""
+    for index, shape_key in extract_shape_keys(node.data.shape_keys):
+
+        node.show_only_shape_key = True
+        node.active_shape_key_index = index
+        shape_key.value = 1.0
+
+        mesh = node.to_mesh(bpy.context.scene,
+                            export_settings['use_mesh_modifiers'],
+                            "RENDER")
+        triangulate_mesh(mesh)
+
+        if has_tangents:
+            mesh.calc_tangents()
+        else:
+            mesh.calc_normals_split()
+
+        surface_to_morph = [None] * len(surfaces)
+
+        for face_index in range(len(mesh.polygons)):
+            face = mesh.polygons[face_index]
+            morph = None
+            surface = None
+            for surf_index, surf in enumerate(surfaces):
+                # todo:
+                # use the `material_to_surface` map from `generate_surfaces`
+                if surf.face_material_index == face.material_index:
+                    if surface_to_morph[surf_index] is None:
+                        surface_to_morph[surf_index] = VerticesArrays()
+                        surface_to_morph[surf_index].vertices = [
+                            None] * len(surf.vertex_data.vertices)
+
+                    morph = surface_to_morph[surf_index]
+                    surface = surf
+                    break
+
+            if morph is not None and surface is not None:
+
+                for loop_id in range(face.loop_total):
+                    loop_index = face.loop_start + loop_id
+                    new_vert = export_vertex_from_loop(
+                        mesh, loop_index, has_tangents, gid_to_bid_map)
+
+                    morph.vertices[
+                        surface.vertex_index_map[loop_index]] = new_vert
+
+        for surf_index, surf in enumerate(surfaces):
+            if surface_to_morph[surf_index] is not None:
+                surf.morph_arrays.append(surface_to_morph[surf_index])
+
+        bpy.data.meshes.remove(mesh)
+
+
+def export_vertex_from_loop(mesh, loop_index, has_tangents, gid_to_bid_map):
```

Comment:
I wonder if it would make sense to have this as a @classmethod of Vertex(). Something like: Vertex.create_from_loop()

Reply:
yeah, it can, but I think it is not appropriate as a constructor,  maybe makes it as a staticmethod of Vertex(),